ast/src/analyzed/mod.rs

```diff
@@ -715,96 +711,296 @@ impl<Expr> Default for SelectedExpressions<Expr> {
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        Box::new(self.left.children().chain(self.right.children()))
+    }
+}
+
+impl<T> PermutationIdentity<T> {
+    pub fn new(
+        id: u64,
+        source: SourceRef,
+        left: SelectedExpressions<T>,
+        right: SelectedExpressions<T>,
+    ) -> Self {
+        Self {
+            id,
+            source,
+            left,
+            right,
+        }
+    }
+}
+
+impl<Expr> TryFrom<Identity<Expr>> for PermutationIdentity<Expr> {
+    type Error = ();
+
+    fn try_from(value: Identity<Expr>) -> Result<Self, Self::Error> {
+        match value {
+            Identity::Permutation(permutation) => Ok(permutation),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct ConnectIdentity<T> {
+    // The ID is globally unique among identitites.
+    pub id: u64,
+    pub source: SourceRef,
+    pub left: SelectedExpressions<T>,
+    pub right: SelectedExpressions<T>,
+}
+
+impl<T> Children<AlgebraicExpression<T>> for ConnectIdentity<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        Box::new(self.left.children_mut().chain(self.right.children_mut()))
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        Box::new(self.left.children().chain(self.right.children()))
+    }
+}
+
+impl<T> ConnectIdentity<T> {
+    pub fn new(
+        id: u64,
+        source: SourceRef,
+        left: SelectedExpressions<T>,
+        right: SelectedExpressions<T>,
+    ) -> Self {
+        Self {
+            id,
+            source,
+            left,
+            right,
+        }
+    }
+}
+
+impl<T> TryFrom<Identity<T>> for ConnectIdentity<T> {
+    type Error = ();
+
+    fn try_from(value: Identity<T>) -> Result<Self, Self::Error> {
+        match value {
+            Identity::Connect(connect) => Ok(connect),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize, JsonSchema, derive_more::Display)]
+pub enum Identity<T> {
+    Polynomial(PolynomialIdentity<T>),
+    Plookup(PlookupIdentity<T>),
+    Permutation(PermutationIdentity<T>),
+    Connect(ConnectIdentity<T>),
+}
+
+impl<T> From<PolynomialIdentity<T>> for Identity<T> {
+    fn from(value: PolynomialIdentity<T>) -> Self {
+        Self::Polynomial(value)
+    }
+}
+
+impl<T> From<PlookupIdentity<T>> for Identity<T> {
+    fn from(value: PlookupIdentity<T>) -> Self {
+        Self::Plookup(value)
+    }
+}
+
+impl<T> From<PermutationIdentity<T>> for Identity<T> {
+    fn from(value: PermutationIdentity<T>) -> Self {
+        Self::Permutation(value)
+    }
+}
+
+impl<T> From<ConnectIdentity<T>> for Identity<T> {
+    fn from(value: ConnectIdentity<T>) -> Self {
+        Self::Connect(value)
+    }
+}
+
+// TODO This is the only version of Identity left.
```

Comment:
@chriseth can I remove this comment?

Reply:
The idea behind this comment was to refactor Identity so that it is only generic over the field, and not about what a selected expressions or an expression / algebraic expression is. If you did this change, then you can remove it.

Reply:
Oh I didnt get that but yes that's exactly what I did :)

Reply:
Can you remove the TODO?

Reply:
done